crmprtd/insert.py

```diff
@@ -8,11 +8,217 @@
 speed and reliability. This phase is common to all networks.
 """
 
+from math import floor
+from sqlalchemy import and_
+import logging
+import time
+from sqlalchemy.exc import IntegrityError
+
+from crmprtd.db_exceptions import UniquenessError, InsertionError
+from pycds import Obs
+
+
+log = logging.getLogger('crmprtd.insert')
+
+
+class DBMetrics(object):
+    '''Keep track of database metrics during the insertion process.
+    '''
+
+    def __init__(self):
+        self.successes = 0
+        self.failures = 0
+
+    def clear(self):
+        self.successes = 0
+        self.failures = 0
+
+
+class Timer(object):
+    def __enter__(self):
+        self.start = time.time()
+        return self
+
+    def __exit__(self, *args):
+        self.end = time.time()
+        self.run_time = self.end - self.start
+
+
+def chunks(list, chunk_size):
+    for i in range(0, len(list), chunk_size):
+        yield list[i:i+chunk_size]
```

Comment:
Hmmm interesting. This does work, but only because Python will (surprisingly) drop the end of an array if you give it an index that is larger than the length of the array.

I suspect that you could do this more elegantly with `itertools.groupby`, but this works, so I'm satisfied to go with this.